From 6bce7561dd461cb9c0c4635b10e4a12f9cc4f9b5 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Thu, 16 Jan 2020 13:41:11 +0100
Subject: [PATCH] UGW_SW-46943: Create /dev/ltq_icc with hotplug

This will tell the user space to create a device /dev/ltq_icc automatically
with the given major and minor numbers. In OpenWrt this is done by
procd, but other Linux distributions can use different hotplug daemons.

The /dev/ltq_icc interface is only needed on the TEP and not on the
InetrAptiv, compile it out on the InterAptiv for security reasons.
---
 drivers/char/ltq_icc/drv_icc.c | 35 ++++++++++++++++++++++++++++++++++-
 1 file changed, 34 insertions(+), 1 deletion(-)

diff --git a/drivers/char/ltq_icc/drv_icc.c b/drivers/char/ltq_icc/drv_icc.c
index c923c4ff2e24..b4907d0a2d95 100644
--- a/drivers/char/ltq_icc/drv_icc.c
+++ b/drivers/char/ltq_icc/drv_icc.c
@@ -91,7 +91,11 @@ typedef struct{
 }mmap_addr_t;
 
 static short icc_major_id = 0;
-char icc_dev_name[] = "ltq_icc";
+#ifdef CONFIG_SOC_TYPE_GRX500_TEP
+static struct device *icc_char_dev;
+static struct class *icc_char_class;
+const char icc_dev_name[] = "ltq_icc";
+#endif
 /*Variable used*/
 #ifdef KTHREAD
 unsigned int g_num;
@@ -568,6 +572,7 @@ unsigned int icc_poll(struct file *file_p, poll_table *wait){
 	return ret;
 }
 
+#ifdef CONFIG_SOC_TYPE_GRX500_TEP
 /*register the char region with the linux*/
 int icc_os_register(void)
 {
@@ -579,6 +584,23 @@ int icc_os_register(void)
 		return ret;
 	} else if (icc_major_id == 0) {
 		icc_major_id = ret;
+
+		icc_char_class = class_create(THIS_MODULE, icc_dev_name);
+		if (IS_ERR(icc_char_class)) {
+			TRACE(ICC, DBG_LEVEL_HIGH, ("failed to create device class\n"));
+			unregister_chrdev(icc_major_id, icc_dev_name);
+			return PTR_ERR(icc_char_class);
+		}
+
+		icc_char_dev = device_create(icc_char_class, NULL,
+					     MKDEV(icc_major_id, 0),
+					     NULL, "%s", icc_dev_name);
+		if (IS_ERR(icc_char_dev)) {
+			TRACE(ICC, DBG_LEVEL_HIGH, ("failed to create device\n"));
+			class_destroy(icc_char_class);
+			unregister_chrdev(icc_major_id, icc_dev_name);
+			return PTR_ERR(icc_char_dev);
+		}
 	}
 	TRACE(ICC, DBG_LEVEL_HIGH, ("Major Id is %d\n", icc_major_id));
    	TRACE(ICC, DBG_LEVEL_HIGH, ("ICC driver registered\n"));
@@ -591,9 +613,20 @@ int icc_os_register(void)
 void icc_os_unregister (void)
 {
 	 /*unregister the driver region completely*/
+	 device_unregister(icc_char_dev);
+	 class_destroy(icc_char_class);
    unregister_chrdev(icc_major_id, icc_dev_name);
 	 TRACE(ICC, DBG_LEVEL_HIGH, ("ICC driver un-registered\n"));
 }
+#else
+static int icc_os_register(void)
+{
+	return 0;
+}
+static void icc_os_unregister(void)
+{
+}
+#endif /* CONFIG_SOC_TYPE_GRX500_TEP */
 
 /*Global Functions*/
 /**
